File: backend/workflow/spawn.go
package workflow

import (
	"context"
	"encoding/json"
	"fmt"
	"strings"
	"time"

	"github.com/enough/enough/backend/agent"
	"github.com/enough/enough/backend/core"
)

func (r *Runtime) executeJob(parent context.Context, phase, key string, opts AgentOptions) AgentResult {
	if opts.Role == "" {
		opts.Role = phase
	}
	if opts.SystemPrompt == "" {
		opts.SystemPrompt = roleTemplate(opts.Role)
	}
	if len(opts.ResponseSchema) > 0 {
		if schema, err := json.MarshalIndent(opts.ResponseSchema, "", "  "); err == nil {
			opts.Prompt += "\n\nRespond with JSON only matching this schema:\n" + string(schema)
		}
	}
	if opts.Role == "audit" || opts.Role == "rule" || opts.Role == "verify" {
		opts.Readonly = true
	}
	if !r.reserveAgent() {
		result := AgentResult{Key: key, Role: opts.Role, OK: false, Error: fmt.Sprintf("workflow exceeded maxTotalAgents=%d", r.maxTotalAgents)}
		r.finishAgent(key, result)
		return result
	}

	ctx, cancel := context.WithCancel(parent)
	r.startAgent(key, phase, opts, cancel)
	result := r.runAgentAttempt(ctx, phase, key, opts)
	if result.Error == "" && len(opts.ResponseSchema) > 0 {
		value, err := parseAndValidateJSON(result.Text, opts.ResponseSchema)
		if err != nil {
			if !r.reserveAgent() {
				result.Error = err.Error()
			} else {
				retry := opts
				retry.Prompt = opts.Prompt + "\n\nYour previous response was invalid: " + err.Error() + "\nRespond with JSON only matching the response schema."
				result = r.runAgentAttempt(ctx, phase, key, retry)
				if result.Error == "" {
					value, err = parseAndValidateJSON(result.Text, opts.ResponseSchema)
				}
				if err != nil {
					result.Error = err.Error()
				} else {
					result.JSON = value
				}
			}
		} else {
			result.JSON = value
		}
	}
	result.OK = result.Error == ""
	r.finishAgent(key, result)
	cancel()
	return result
}

func (r *Runtime) runAgentAttempt(ctx context.Context, phase, key string, opts AgentOptions) AgentResult {
	return r.agentRunner(ctx, phase, key, opts, func(event core.Event) {
		r.handleWorkerEvent(phase, key, opts.Role, event)
	})
}

func (r *Runtime) defaultAgentRunner(ctx context.Context, phase, key string, opts AgentOptions, emit func(core.Event)) AgentResult {
	worker := agent.RunWorkflowAgent(ctx, r.cfg, r.workDir, agent.WorkflowAgentOptions{
		Prompt: opts.Prompt, SystemPrompt: opts.SystemPrompt, Tools: opts.Tools,
		Model: opts.Model, MaxTurns: opts.MaxTurns, Readonly: opts.Readonly,
	}, emit)
	return AgentResult{
		Key: key, Role: opts.Role, Text: worker.Text, Error: worker.Error,
		TokensUsed: worker.TokensUsed, TurnCount: worker.TurnCount,
	}
}

func (r *Runtime) reserveAgent() bool {
	r.mu.Lock()
	defer r.mu.Unlock()
	if r.maxTotalAgents > 0 && r.totalAgents >= r.maxTotalAgents {
		return false
	}
	r.totalAgents++
	return true
}

func (r *Runtime) startAgent(key, phase string, opts AgentOptions, cancel context.CancelFunc) {
	now := time.Now()
	r.mu.Lock()
	r.active[key] = &jobControl{cancel: cancel}
	r.ensurePhaseLocked(phase)
	s := r.snapshot.Agents[key]
	s.Key, s.Phase, s.Role, s.Prompt = key, phase, defaultRole(opts.Role), opts.Prompt
	s.Status, s.StartedAt = "running", now
	r.snapshot.Agents[key] = s
	r.recountLocked()
	emit := r.emit
	workflowID := r.snapshot.ID
	r.mu.Unlock()
	if emit != nil {
		emit(core.Event{Kind: core.EventWorkflowAgentStart, Data: core.WorkflowAgentEvent{
			WorkflowID: workflowID, Phase: phase, Key: key, Role: opts.Role,
			Status: "running", Prompt: opts.Prompt,
		}})
	}
	r.emitRun(core.EventWorkflowPhase, phase)
}

func (r *Runtime) finishAgent(key string, result AgentResult) {
	now := time.Now()
	r.mu.Lock()
	delete(r.active, key)
	s := r.snapshot.Agents[key]
	s.Status = "done"
	if !result.OK {
		s.Status = "failed"
	}
	s.Result, s.JSON, s.Error = result.Text, result.JSON, result.Error
	s.Tokens, s.Turns, s.EndedAt = result.TokensUsed, result.TurnCount, now
	r.snapshot.Agents[key] = s
	cancelledForPause := r.paused && strings.Contains(strings.ToLower(result.Error), "context canceled")
	if !isQuotaError(result.Error) && !r.restartRequested[key] && !cancelledForPause {
		r.completed[key] = result
	}
	r.recountLocked()
	emit := r.emit
	workflowID := r.snapshot.ID
	r.mu.Unlock()
	r.persist()
	if emit != nil {
		emit(core.Event{Kind: core.EventWorkflowAgentEnd, Data: core.WorkflowAgentEvent{
			WorkflowID: workflowID, Phase: s.Phase, Key: key, Role: result.Role,
			Status: s.Status, Result: result.Text, JSON: result.JSON, Error: result.Error,
			Tokens: result.TokensUsed, Turns: result.TurnCount,
		}})
	}
	r.emitRun(core.EventWorkflowPhase, s.Phase)
}

func (r *Runtime) handleWorkerEvent(phase, key, role string, event core.Event) {
	tool, ok := event.Data.(core.ToolCallEvent)
	if !ok {
		return
	}
	r.mu.Lock()
	s := r.snapshot.Agents[key]
	if event.Kind == core.EventToolStart {
		s.LastTools = append(s.LastTools, tool.Name)
		if len(s.LastTools) > 12 {
			s.LastTools = s.LastTools[len(s.LastTools)-12:]
		}
	}
	r.snapshot.Agents[key] = s
	emit := r.emit
	workflowID := r.snapshot.ID
	r.mu.Unlock()
	if emit != nil {
		emit(core.Event{Kind: core.EventWorkflowAgentDelta, Data: core.WorkflowAgentEvent{
			WorkflowID: workflowID, Phase: phase, Key: key, Role: role,
			Status: "running", Tool: tool,
		}})
	}
}

func defaultRole(role string) string {
	role = strings.TrimSpace(role)
	if role == "" {
		return "agent"
	}
	return role
}

File: backend/workflow/spawn.ts
// PORT: backend/workflow/spawn.go

import { type AgentOptions, type AgentResult, type AgentSnapshot } from "./types";
import { Runtime } from "./runtime";
import { parseAndValidateJSON } from "./schema";
import { roleTemplate } from "./roles";
import { RunWorkflowAgent } from "../agent/workflow_worker";
import {
  event_workflow_agent_start,
  event_workflow_agent_end,
  event_workflow_agent_delta,
  event_workflow_phase,
  event_tool_start,
  type event,
  type tool_call_event,
} from "../core/events";
import { isQuotaError } from "./pool";
import { Effect } from "effect";

export async function executeJob(
  this: Runtime,
  parent: AbortSignal,
  phase: string,
  key: string,
  opts: AgentOptions
): Promise<AgentResult> {
  if (!opts.role) {
    opts.role = phase;
  }
  if (!opts.systemPrompt) {
    opts.systemPrompt = roleTemplate(opts.role);
  }
  if (opts.responseSchema && Object.keys(opts.responseSchema).length > 0) {
    opts.prompt += "\n\nRespond with JSON only matching this schema:\n" + JSON.stringify(opts.responseSchema, null, "  ");
  }
  if (opts.role === "audit" || opts.role === "rule" || opts.role === "verify") {
    opts.readonly = true;
  }

  if (!this.reserveAgent()) {
    const result: AgentResult = {
      key,
      role: opts.role,
      ok: false,
      text: "",
      error: `workflow exceeded maxTotalAgents=${this.maxTotalAgents}`,
    };
    this.finishAgent(key, result);
    return result;
  }

  // Create child controller/signal to handle cancellation
  const controller = new AbortController();
  const onParentAbort = () => controller.abort();
  if (parent.aborted) {
    onParentAbort();
  } else {
    parent.addEventListener("abort", onParentAbort);
  }

  this.startAgent(key, phase, opts, () => controller.abort());

  let result = await this.runAgentAttempt(controller.signal, phase, key, opts);

  if (!result.error && opts.responseSchema && Object.keys(opts.responseSchema).length > 0) {
    try {
      const parsedVal = await Effect.runPromise(parseAndValidateJSON(result.text, opts.responseSchema));
      result.json = parsedVal;
    } catch (err: any) {
      if (!this.reserveAgent()) {
        result.error = err.message || String(err);
      } else {
        const retryOpts = { ...opts };
        retryOpts.prompt = opts.prompt + "\n\nYour previous response was invalid: " + (err.message || String(err)) + "\nRespond with JSON only matching the response schema.";
        result = await this.runAgentAttempt(controller.signal, phase, key, retryOpts);
        if (!result.error) {
          try {
            const parsedVal = await Effect.runPromise(parseAndValidateJSON(result.text, opts.responseSchema));
            result.json = parsedVal;
          } catch (retryErr: any) {
            result.error = retryErr.message || String(retryErr);
          }
        }
      }
    }
  }

  result.ok = !result.error;
  this.finishAgent(key, result);
  
  parent.removeEventListener("abort", onParentAbort);
  controller.abort();
  return result;
}

export async function runAgentAttempt(
  this: Runtime,
  ctx: AbortSignal,
  phase: string,
  key: string,
  opts: AgentOptions
): Promise<AgentResult> {
  return this.agentRunner(ctx, phase, key, opts, (ev: event) => {
    this.handleWorkerEvent(phase, key, opts.role ?? "", ev);
  });
}

export async function defaultAgentRunner(
  this: Runtime,
  ctx: AbortSignal,
  phase: string,
  key: string,
  opts: AgentOptions,
  emit: (event: event) => void
): Promise<AgentResult> {
  const result = await RunWorkflowAgent(
    ctx,
    this.cfg,
    this.workDir,
    {
      Prompt: opts.prompt,
      SystemPrompt: opts.systemPrompt ?? "",
      Tools: opts.tools ?? [],
      Model: opts.model ?? "",
      MaxTurns: opts.maxTurns ?? 0,
      Readonly: !!opts.readonly,
    },
    emit
  );
  return {
    key,
    role: opts.role,
    ok: !result.Error,
    text: result.Text,
    error: result.Error,
    tokensUsed: result.TokensUsed,
    turnCount: result.TurnCount,
  };
}

export function reserveAgent(this: Runtime): boolean {
  if (this.maxTotalAgents > 0 && this.totalAgents >= this.maxTotalAgents) {
    return false;
  }
  this.totalAgents++;
  return true;
}

export function startAgent(
  this: Runtime,
  key: string,
  phase: string,
  opts: AgentOptions,
  cancel: () => void
): void {
  const now = new Date();
  this.active[key] = { cancel };
  this.ensurePhaseLocked(phase);

  const s = this.snapshot.agents[key] || {
    key: "",
    phase: "",
    role: "",
    status: "",
    prompt: "",
  };
  s.key = key;
  s.phase = phase;
  s.role = defaultRole(opts.role ?? "");
  s.prompt = opts.prompt;
  s.status = "running";
  s.startedAt = now;
  this.snapshot.agents[key] = s;

  this.recountLocked();
  const emit = this.emit;
  const workflowID = this.snapshot.id;
  if (emit) {
    emit({
      kind: event_workflow_agent_start,
      data: {
        workflow_id: workflowID,
        phase,
        key,
        role: opts.role ?? "",
        status: "running",
        prompt: opts.prompt,
      },
    });
  }
  this.emitRun(event_workflow_phase, phase);
}

export function finishAgent(this: Runtime, key: string, result: AgentResult): void {
  const now = new Date();
  delete this.active[key];

  const s = this.snapshot.agents[key] || {
    key: "",
    phase: "",
    role: "",
    status: "",
    prompt: "",
  };
  s.status = "done";
  if (!result.ok) {
    s.status = "failed";
  }
  s.result = result.text;
  s.json = result.json;
  s.error = result.error ?? "";
  s.tokens = result.tokensUsed;
  s.turns = result.turnCount;
  s.endedAt = now;
  this.snapshot.agents[key] = s;

  const cancelledForPause = this.paused && (result.error ?? "").toLowerCase().includes("context canceled");
  if (!isQuotaError(result.error ?? "") && !this.restartRequested[key] && !cancelledForPause) {
    this.completed[key] = result;
  }

  this.recountLocked();
  const emit = this.emit;
  const workflowID = this.snapshot.id;
  this.persist();

  if (emit) {
    emit({
      kind: event_workflow_agent_end,
      data: {
        workflow_id: workflowID,
        phase: s.phase,
        key,
        role: result.role ?? "",
        status: s.status,
        result: result.text,
        json: result.json,
        error: result.error ?? "",
        tokens: result.tokensUsed ?? 0,
        turns: result.turnCount ?? 0,
      },
    });
  }
  this.emitRun(event_workflow_phase, s.phase);
}

export function handleWorkerEvent(
  this: Runtime,
  phase: string,
  key: string,
  role: string,
  event: event
): void {
  if (event.kind !== event_tool_start) {
    return;
  }
  const tool = event.data as tool_call_event;
  if (!tool) {
    return;
  }

  const s = this.snapshot.agents[key];
  if (!s) return;
  if (!s.lastTools) {
    s.lastTools = [];
  }
  s.lastTools.push(tool.name);
  if (s.lastTools.length > 12) {
    s.lastTools = s.lastTools.slice(s.lastTools.length - 12);
  }
  this.snapshot.agents[key] = s;

  const emit = this.emit;
  const workflowID = this.snapshot.id;
  if (emit) {
    emit({
      kind: event_workflow_agent_delta,
      data: {
        workflow_id: workflowID,
        phase,
        key,
        role,
        status: "running",
        tool,
      },
    });
  }
}

export function defaultRole(role: string): string {
  role = role.trim();
  if (role === "") {
    return "agent";
  }
  return role;
}

// Attach prototype methods
const proto = Runtime.prototype as any;
proto.executeJob = executeJob;
proto.runAgentAttempt = runAgentAttempt;
proto.defaultAgentRunner = defaultAgentRunner;
proto.reserveAgent = reserveAgent;
proto.startAgent = startAgent;
proto.finishAgent = finishAgent;
proto.handleWorkerEvent = handleWorkerEvent;

/*
PORT STATUS
source path: backend/workflow/spawn.go
source lines: 178
draft lines: 275
confidence: high
status: phase_b_compile
*/
